### PR TITLE
function development turnElement

### PR DESCRIPTION
**Pull request by PashaBeshchuk**

### CODE REVIEW

File: classes.js [hunk 1]
```diff
@@ -0,0 +1 @@
+class
```
Comment (zbstof) [inline classes.js]: Why did you add this file?
Please remove

File: tetris.js [hunk 2]
```diff
@@ -74,7 +74,9 @@ module.exports = {
 			}
 			
 			for(let i = 0; i < arrCoordinates_X.length; i++){
-				if(field[coordinates_Y[i]][coordinates_X[i]] == true){
```
Comment (zbstof) [inline tetris.js]: This function has several problems:
1. It never returns false, only undedefined - please make it so it only returns true or false, nothing else
2. it's not clear what is the type of elem? Is is single element? Is is array of elements? Rename parameter and add commentary above the function that explains it's purpose

File: tetris.js [hunk 3]
```diff
@@ -100,7 +102,11 @@ module.exports = {
 					}
 					newCoordenats.push(obj)
 				}
-				return newCoordenats
+				if(checkThatTheFieldIsFree(newCoordenats)){
+					return localTetromino
+				}else{
+					return newCoordenats
```
Comment (zbstof) [inline tetris.js]: google the translation of the word "координаты", correct the spelling of this variable

File: tetris.js [hunk 4]
```diff
@@ -100,7 +102,11 @@ module.exports = {
 					}
 					newCoordenats.push(obj)
 				}
-				return newCoordenats
+				if(checkThatTheFieldIsFree(newCoordenats)){
+					return localTetromino
+				}else{
```
Comment (zbstof) [inline tetris.js]: fix formatting.
VSCode will do this for you automatically

File: tetris.js [hunk 5]
```diff
@@ -229,10 +247,15 @@ module.exports = {
 		return newArr;
 	},
 	
-	buildTetromino: function(stringOfCoordinates){
-		let math = stringOfCoordinates.split("\n")
-		let arrayOfCoordinates = []
+	getCoordinates: function(stringOfCoordinates){
+		let math
```
Comment (zbstof) [inline tetris.js]: what is "math"? Please find more suitable variable name

File: tetris.js [hunk 6]
```diff
@@ -229,10 +247,15 @@ module.exports = {
...
 		let quantityElem = 0;
+		let arrayOfCoordinates = []
```
Comment (zbstof) [inline tetris.js]: You created this empty array and are returning it on line 282. But you are not updating it anywhere - looks like a bug

File: tetris.js [hunk 7]
```diff
@@ -241,37 +264,46 @@ module.exports = {
...
 							obj.x = elem_X
 							obj.y = elem_Y
 							quantityElem++;
 							
-							arrayOfCoordinates.push(obj)
+							arrayOfCoordinates.push(obj);
 							
 						}
-						elem_X++
+						elem_X++;
```
Comment (zbstof) [inline tetris.js]: this is not element value, please find more suitable variable name

File: tetris.js [hunk 8]
```diff
@@ -241,37 +264,46 @@ module.exports = {
...
 					}
 					
 				}
-				elem_Y++
+				elem_Y++;
```
Comment (zbstof) [inline tetris.js]: this is not element value, please find more suitable variable name

File: tetris.js [hunk 9]
```diff
@@ -241,37 +264,46 @@ module.exports = {
...
 			}
 			
 		}
+		return arrayOfCoordinates;
+	},
+	
+	buildTetromino: function(stringOfCoordinates){
+		let math = stringOfCoordinates.split("\n");
+		let arrayOfCoordinates = [];
+		module.exports.getCoordinates(math);
```
Comment (zbstof) [inline tetris.js]: why are you invoking this function here but not doing anything with results?

Comment (zbstof) [reply tetris.js]: Why are you using function via module exports? Please invoke the function directly. And the next line as well.

File: tetris.js [hunk 10]
```diff
@@ -241,37 +264,46 @@ module.exports = {
...
+		module.exports.getCoordinates(math).forEach(function(elem){
```
Comment (zbstof) [inline tetris.js]: here

File: tetris.js (hunk 4 above)
Comment (PashaBeshchuk) [reply tetris.js]: formatting is correct, there is no error in my file

Comment (zbstof) [reply tetris.js]: No, there should be a space after '}' and the space '{' everywhere.
Please insteall VScode - it will fix this

File: tetris.js (hunk 6 above)
Comment (PashaBeshchuk) [reply tetris.js]: I work with this array

File: tetris.js (hunk 9 above)
Comment (PashaBeshchuk) [reply tetris.js]: "why are you invoking this function here but not doing anything with results? "- I do not understand  this.
I work with function result.

"Why are you using function via module exports? Please invoke the function directly. And the next line as well." - I don't know how to do it

File: tetris.js [hunk 11]
```diff
@@ -229,49 +242,63 @@ module.exports = {
...
 				
-				for(let j = 0, elem_X = 0; j < math[i].length; j++){
+				for(let j = 0, axisElement_X = 0; j < fieldArray[i].length; j++){
 					let obj = {}
-					if(math[i][j]!= false){
-						if(math[i][j] == "X"){
-							console.log(math[i][j])
-							obj.x = elem_X
-							obj.y = elem_Y
+					if(fieldArray[i][j]!= false){
```
Comment (zbstof) [inline tetris.js]: DOn't compare strings with booleans, please

File: tetris.js [hunk 12]
```diff
@@ -229,49 +242,63 @@ module.exports = {
...
-	buildTetromino: function(stringOfCoordinates){
-		let math = stringOfCoordinates.split("\n")
-		let arrayOfCoordinates = []
+	getCoordinates: function(stringOfCoordinates){
+		let fieldArray
 		let quantityElem = 0;
+		let arrayOfCoordinates = []
+		if(typeof(stringOfCoordinates) == "string"){
+			fieldArray = stringOfCoordinates.split("\n")
+		}else{
```
Comment (zbstof) [inline tetris.js]: is it possible for stringOfCoordinates to not be of type "string"?
When?

File: tetris.js [hunk 13]
```diff
@@ -10,6 +10,17 @@ module.exports = {
 		let a = -1;
 		let localTetromino = []
 		
+		//check that the field is free to rotate
```
Comment (zbstof) [inline tetris.js]: we never rotate field

File: tetris.js [hunk 14]
```diff
@@ -10,6 +10,17 @@ module.exports = {
 		let a = -1;
 		let localTetromino = []
 		
+		//check that the field is free to rotate	
+		function checkThatTheFieldIsFree(elem){
```
Comment (zbstof) [inline tetris.js]: my point #2 is not fulfilled

File: tetris.js [hunk 15]
```diff
@@ -10,6 +10,17 @@ module.exports = {
...
+			for(let i of elem){
+				//undefined - if this part of the field is missing
```
Comment (zbstof) [inline tetris.js]: explain further, in which case the field will be missing

File: tetris.js [hunk 16]
```diff
@@ -10,6 +10,17 @@ module.exports = {
 		let a = -1;
 		let localTetromino = []
 		
+		//check that the field is free to rotate	
+		function checkThatTheFieldIsFree(elem){
+			for(let i of elem){
```
Comment (zbstof) [inline tetris.js]: i is for index - this is definitely not index - please rename

File: tetris.js (hunk 4 above)
Comment (zbstof) [reply tetris.js]: https://stackoverflow.com/a/29973358/362721

File: tetris.js [hunk 17]
```diff
@@ -229,49 +242,63 @@ module.exports = {
...
+		let arrayOfCoordinates = []
+		if(typeof(stringOfCoordinates) == "string"){
+			fieldArray = stringOfCoordinates.split("\n")
+		}else{
+			fieldArray = stringOfCoordinates
+		}
 		
-		for(let i = 0, elem_Y = 0; i < math.length; i++){
-			if(math[i]!= false){
+		for(let i = 0, axisElement_Y = 0; i < fieldArray.length; i++){
```
Comment (zbstof) [inline tetris.js]: this is not element, but index. Element is for elements of some list-data-structure, like array

File: tetris.js [hunk 18]
```diff
@@ -229,49 +242,63 @@ module.exports = {
...
+			if(fieldArray[i]!= false){
 				
-				for(let j = 0, elem_X = 0; j < math[i].length; j++){
+				for(let j = 0, axisElement_X = 0; j < fieldArray[i].length; j++){
```
Comment (zbstof) [inline tetris.js]: and here

File: tetris.js [hunk 19]
```diff
@@ -229,49 +242,63 @@ module.exports = {
...
 			}
 			
 		}
+		return arrayOfCoordinates;
+	},
+	
+	buildTetromino: function(stringOfCoordinates){
+		let fieldArray = stringOfCoordinates.split("\n");
+		let arrayOfCoordinates = [];
+		module.exports.getCoordinates(fieldArray);
```
Comment (zbstof) [inline tetris.js]: 1. where do you use the result of this function invocation? exact line, please
2. Are you changing fieldArray inside getCoordinates?

File: tetris.js [hunk 20]
```diff
@@ -229,49 +242,63 @@ module.exports = {
...
+		module.exports.getCoordinates(fieldArray).forEach(function(elem){
```
Comment (zbstof) [inline tetris.js]: remove "module.exports." from this invocation

File: tetris.js (hunk 6 above)
Comment (zbstof) [reply tetris.js]: Ok

**Comment by zbstof** [comment]
Format all of these files - it's hard to keep track of the nested functions without proper indentation

File: tetris.js (hunk 20 above)
Comment (PashaBeshchuk) [reply tetris.js]: it won't work like that

File: tetris.js (hunk 19 above)
Comment (PashaBeshchuk) [reply tetris.js]: line 279, 284, 285...

File: tetris.js (hunk 15 above)
Comment (PashaBeshchuk) [reply tetris.js]: During tests, if at turning the tetramino will go out of the field

File: tetris.js (hunk 11 above)
Comment (PashaBeshchuk) [reply tetris.js]: why can't i use booleans?
Everything works, I do not see the error

File: tetris.js (hunk 12 above)
Comment (PashaBeshchuk) [reply tetris.js]: corrected

File: tetris.js (hunk 11 above)
Comment (zbstof) [reply tetris.js]: https://stackoverflow.com/a/359509/362721

File: tetris.js (hunk 9 above)
Comment (zbstof) [reply tetris.js]: > I work with function result.
Please specify exact line where you use result of function invoked on line 288

Comment (zbstof) [reply tetris.js]: > I don't know how to do it
OK

File: tetris.js (hunk 12 above)
Comment (zbstof) [reply tetris.js]: No it's not - otherwise this comment would be marked as "Outdated". If you disagree - attach screenshot

File: tetris.js (hunk 15 above)
Comment (zbstof) [reply tetris.js]: Why not check for negative coordinate instead?

File: tetris.js (hunk 19 above)
Comment (zbstof) [reply tetris.js]: 1. Wrong
2. Please answer the second question

File: tetris.js (hunk 20 above)
Comment (zbstof) [reply tetris.js]: Copy the exact error you are getting here, please

File: tetris.js (hunk 19 above)
Comment (zbstof) [reply tetris.js]: 1. Also 279 is not even line of code

Comment (zbstof) [reply tetris.js]: 1. Hint: are you using the result of the invocation in line 283?

File: tetris.js (hunk 20 above)
Comment (PashaBeshchuk) [reply tetris.js]: ReferenceError: getCoordinates is not defined

File: tetris.js (hunk 19 above)
Comment (PashaBeshchuk) [reply tetris.js]: I did

File: tetris.js (hunk 15 above)
Comment (PashaBeshchuk) [reply tetris.js]: line 421 in file test-tetris.js have test, L-tetromino too low to turn.  
When tetramino rotates and  it cannot turn because the field is absent, then the field along the field[i.x] axis or field[i.y] axis === undefined.

File: tetris.js (hunk 17 above)
Comment (PashaBeshchuk) [reply tetris.js]: I did

File: tetris.js (hunk 15 above)
Comment (zbstof) [reply tetris.js]: 1. Rewrite allItemsOnTheField so it would return boolean (true or false).
2. Rewrite checkThatTheFieldIsFree to use allItemsOnTheField to check if tetramino is on the field

File: tetris.js [hunk 21]
```diff
@@ -9,6 +9,16 @@ module.exports = {
...
+			for(let i of elem){
+				//undefined - if this part of the field is missing
+				if(field[i.x] == undefined || field[i.y] == undefined || field[i.y][i.x] == true){
+					return true
+				}
+			}
+			return false
+		}
 		
 		tetromino.forEach(function(elem){
```
Comment (zbstof) [inline tetris.js]: Rename all variables called 'elem' to something more suitable for the purpose of the each variable

File: tetris.js (hunk 11 above)
Comment (PashaBeshchuk) [reply tetris.js]: //typeof(fieldArray[i][j] ==="string"
I checked what kind of object it is "string"
if(fieldArray[i][j]!== "string"){

**Comment by zbstof** [comment]
When the type of this expression will not be "string"?

PashaBeshchuk <notifications@github.com> 于 2019年5月6日周一 21:26写道：

File: tetris.js (hunk 11 above)
Comment (zbstof) [reply tetris.js]: Question: will you ever encounter situation when fieldArray[i][j] is not of type string?

File: tetris.js [hunk 22]
```diff
@@ -1,185 +1,207 @@
...
+			for (let i of turningPoint) {
 				i.x = arrCoordinates_X[1];
 				i.y = arrCoordinates_Y[0] + a;
 				a++;
 			}
-			let newPosition = allItemsOnTheField(turningPoint)
-			if(checkThatTheFieldIsFree(newPosition)){
-				return localTetromino
+
+			 if(checkThatTheFieldIsFree(turningPoint)){
```
Comment (zbstof) [inline tetris.js]: please re-format the file

File: tetris.js [hunk 23]
```diff
@@ -1,185 +1,207 @@
...
+				return turningPoint
```
Comment (zbstof) [inline tetris.js]: I don't like the name of this variable - point means it's one coordinate, but here we have the collection of the coordinates. And I don't see anyone turning around this collection, so it's not really turning, right?

File: tetris.js [hunk 24]
```diff
@@ -1,185 +1,207 @@
...
+						obj.x = tetromino[i].x - 1;
+						obj.y = tetromino[i].y + 1;
+					} else if (j === 3) {
+						obj.x = tetromino[i].x - 1;
+						obj.y = tetromino[i].y + 1;
 					}
-					newCoordenats.push(obj)
+					newCoordinates.push(obj)
+				}
+				if(checkThatTheFieldIsFree(newCoordinates)){
```
Comment (zbstof) [inline tetris.js]: Please reformat file

File: tetris.js [hunk 25]
```diff
@@ -197,85 +219,92 @@ module.exports = {
...
 		let quantityElem = 0;
-		
-		for(let i = 0, elem_Y = 0; i < math.length; i++){
-			if(math[i]!= false){
-				
-				for(let j = 0, elem_X = 0; j < math[i].length; j++){
+		let arrayOfCoordinates = []
+
+		for (let i = 0, axisIndex_Y = 0; i < fieldArray.length; i++) {
+			if (fieldArray[i] !== "") {
```
Comment (zbstof) [inline tetris.js]: 👍

File: tetris.js [hunk 26]
```diff
@@ -197,85 +219,92 @@ module.exports = {
...
+
+				for (let j = 0, axisIndex_X = 0; j < fieldArray[i].length; j++) {
 					let obj = {}
-					if(math[i][j]!= false){
-						if(math[i][j] == "X"){
-							console.log(math[i][j])
-							obj.x = elem_X
-							obj.y = elem_Y
+					if (fieldArray[i][j]=== "-" || fieldArray[i][j]=== "X") {
+						if (fieldArray[i][j] === "X") {
```
Comment (zbstof) [inline tetris.js]: you have 2 nested checks 
`fieldArray[i][j] === "X"`

Please rewrite so you are only checking this once

File: classes.js (hunk 1 above)
Comment (zbstof) [reply classes.js]: do it, please
